File: kill_course.py
import requests
import lxml.etree
import time
import json
import re
import hdu_login

def get_course_sort_id_true(list):
    pattern = r"queryCourse\(this,'(\d+)'"
    pattern1 = r"queryCourse\(this,'(?:[^']*)','(\w+)'"
    extracted_texts = {}
    for text in list:
        match = re.search(pattern, text)
        match1 = re.search(pattern1, text)
        if match:
            extracted_text = match1.group(1)
            id = match.group(1)
            extracted_texts[id] = extracted_text
    return extracted_texts

def get_xkkz_id(html):
    global ccdm,bh_id,jg_id,xsbj,xz,mzm,xslbdm,xbm,zyfx_id,xqh_id
    tree = lxml.etree.HTML(html)
    course_sort_id = tree.xpath('//a[@role="tab"]/@onclick')
    ccdm=tree.xpath('//input[@name="ccdm"]/@value')[0]
    bh_id=tree.xpath('//input[@name="bh_id"]/@value')[0]
    jg_id=tree.xpath('//input[@name="jg_id_1"]/@value')[0]
    xsbj=tree.xpath('//input[@name="xsbj"]/@value')[0]
    xz=tree.xpath('//input[@name="xz"]/@value')[0]
    mzm=tree.xpath('//input[@name="mzm"]/@value')[0]
    xslbdm=tree.xpath('//input[@name="xslbdm"]/@value')[0]
    xbm=tree.xpath('//input[@name="xbm"]/@value')[0]
    zyfx_id=tree.xpath('//input[@name="zyfx_id"]/@value')[0]
    xqh_id=tree.xpath('//input[@name="xqh_id"]/@value')[0]
    return get_course_sort_id_true(course_sort_id)


def get_do_jxb_id(kch_id,jxb_id,kklxdm):
    url="https://newjw.hdu.edu.cn/jwglxt/xsxk/zzxkyzbjk_cxJxbWithKchZzxkYzb.html"
    data={
        "bklx_id" : "0",
        "njdm_id" : njdm_id,
        "xkxnm" : XueNian,
        "xkxqm" : xqm,
        "kklxdm" : kklxdm,
        "kch_id" : kch_id,
        "xkkz_id" : xkkz_id[kklxdm],
        'xsbj': xsbj, 
        'ccdm': ccdm,
        'xz': xz, 
        'mzm': mzm, 
        'xslbdm': xslbdm, 
        'xbm': xbm, 
        'bh_id': bh_id,
        'zyfx_id': zyfx_id,
        'jg_id': jg_id,
        'xqh_id': xqh_id, 
    }
    response=session.post(url,data=data)
    response=response.json()
    for i in response:
        if i["jxb_id"]==jxb_id:
            return i["do_jxb_id"]

def choose(jxb_ids,kch_id,kklxdm,jxbzc):
    url="https://newjw.hdu.edu.cn/jwglxt/xsxk/zzxkyzbjk_xkBcZyZzxkYzb.html"
    data={
        "jxb_ids" : jxb_ids,
        "kch_id" : kch_id,
        "qz" : "0",
    }
    if kklxdm=="01":
        data["njdm_id"]="20"+jxbzc[0:2]
        data["zyh_id"]=jxbzc[2:6]
    response=session.post(url,data=data)
    print(response.text)

def not_choose(jxb_ids,kch_id):
    url="https://newjw.hdu.edu.cn/jwglxt/xsxk/zzxkyzb_tuikBcZzxkYzb.html"
    data={
        "jxb_ids" : jxb_ids,
        "kch_id" : kch_id,
    }
    response=session.post(url,data=data)
    print(response.text)

if __name__ == "__main__":
    with open('config.json', 'r') as f:
        config = json.load(f)

    XueNian=config["time"]["XueNian"]
    xnmc="{}-{}".format(XueNian,str(int(XueNian)+1))
    xqmc=config["time"]["XueQi"]
    if xqmc=='1':
        xqm="3"
    else:
        xqm="12"

    njdm_id="20"+config["login"]["username"][0:2]
    course=config["course"]

    session=requests.Session()
    session=hdu_login.login(session,config["login"]["username"],config["login"]["password"])

    xk_url="https://newjw.hdu.edu.cn/jwglxt/xsxk/zzxkyzb_cxZzxkYzbIndex.html?gnmkdm=N253512&layout=default"
    response=session.get(xk_url)
    print("xkkz_id获取")
    time.sleep(0.5)
    xkkz_id=get_xkkz_id(response.text)
    with open("hdu_xkkz_id.json","w") as f:
        json.dump(xkkz_id,f,ensure_ascii=False,indent=2)

    with open('hdu_kc.json', 'r') as f:
        courses = json.load(f)

    for i in course:
        for j in range(len(courses)):
            if courses[j]["jxbmc"]==i:
                kch_id=courses[j]["kch_id"]
                jxb_id=courses[j]["jxb_id"]
                jxbzc=courses[j]["jxbzc"]
                kklxmc=courses[j]["kklxmc"]
                break

        if kklxmc=="主修课程":
            kklxdm="01"
        elif kklxmc=="通识选修课":
            kklxdm="10"
        elif kklxmc=="体育分项":
            kklxdm="05"
        elif kklxmc=="特殊课程":
            kklxdm="09"
        
        do_jxb_id=get_do_jxb_id(kch_id,jxb_id,kklxdm)
        print("do_jxb_id获取")
        time.sleep(0.5)
        if course[i]=="1":
            choose(do_jxb_id,kch_id,kklxdm,jxbzc)
        else:
            not_choose(do_jxb_id,kch_id)
        time.sleep(3)

File: killCourse.go
package main

import (
	"HDU-KillCourse/client"
	"HDU-KillCourse/config"
	"HDU-KillCourse/log"
	"context"
	"encoding/json"
	"errors"
	"os"
	"time"
)

// killcourse完成
var channel = make(chan string)

// GetDoJxbId 获取doJxbId
func GetDoJxbId(c *client.Client, KchId string, JxbId string, Kklxdm string, NjdmId string, XueNian string, Xqm string) (string, error) {
	// 检查c.ClientBodyConfig是否为nil，测试用
	if c.ClientBodyConfig == nil {
		return "", errors.New("ClientBodyConfig未初始化")
	}

	// 设置请求参数
	req := &client.GetDoJxbIdReq{
		BklxID: "0",
		NjdmID: NjdmId,
		Xkxnm:  XueNian,
		Xkxqm:  Xqm,
		Kklxdm: Kklxdm,
		KchID:  KchId,
		XkkzID: c.ClientBodyConfig.XkkzId[Kklxdm],
		Xsbj:   c.ClientBodyConfig.Xsbj,
		Ccdm:   c.ClientBodyConfig.Ccdm,
		Xz:     c.ClientBodyConfig.Xz,
		Mzm:    c.ClientBodyConfig.Mzm,
		Xslbdm: c.ClientBodyConfig.Xslbdm,
		Xbm:    c.ClientBodyConfig.Xbm,
		BhID:   c.ClientBodyConfig.BhId,
		ZyfxID: c.ClientBodyConfig.ZyfxId,
		JgID:   c.ClientBodyConfig.JgId,
		XqhID:  c.ClientBodyConfig.XqhId,
	}

	// 发送请求
	resp, err := c.GetDoJxbId(req)
	if err != nil {
		return "", err
	}

	// 解析doJxbId
	for _, v := range resp {
		if v.JxbID == JxbId {
			return v.DoJxbID, nil
		}
	}

	return "", errors.New("doJxbId不存在")

}

// SelectCourse 选课
func SelectCourse(c *client.Client, JxbIds string, KchId string, Kklxdm string, Jxbzc string) error {
	// 设置请求参数
	req := &client.SelectCourseReq{
		JxbIDs: JxbIds,
		KchID:  KchId,
		Qz:     "0",
	}

	// 若为主修课程
	if Kklxdm == "01" {
		req.NjdmID = "20" + Jxbzc[0:2]
		req.ZyhID = Jxbzc[2:6]
	}

	// 发送请求
	result, err := c.SelectCourse(req)
	if err != nil {
		return err
	}

	if result.Flag == "1" {
		log.Info("选课成功")
	} else if result.Flag == "0" {
		log.Error("选课失败: ", result.Msg)
	} else {
		log.Error("选课失败: 人数可能已满", result)
	}

	return nil
}

// CancelCourse 退课
func CancelCourse(c *client.Client, JxbIds string, KchId string, XueNian string, Xqm string) error {
	// 设置请求参数
	req := &client.CancelCourseReq{
		JxbIDs: JxbIds,
		KchID:  KchId,
		Xkxnm:  XueNian,
		Xkxqm:  Xqm,
	}

	// 发送请求
	result, err := c.CancelCourse(req)
	if err != nil {
		return err
	}

	if result == "\"1\"" {
		log.Info("退课成功(可能？)")
	} else {
		log.Error("退课失败：", result)
	}

	return nil
}

// HandleCourse 处理课程
func HandleCourse(c *client.Client, cfg *config.Config, course *config.Course, CourseName string, SelectFlag string) error {
	for _, v := range course.Items {
		if v.Jxbmc == CourseName {
			// 更改Kklxdm
			Kklxdm := v.Kklxmc
			if Kklxdm == "主修课程" {
				Kklxdm = "01"
			} else if Kklxdm == "通识选修课" {
				Kklxdm = "10"
			} else if Kklxdm == "体育分项" {
				Kklxdm = "05"
			} else if Kklxdm == "特殊课程" {
				Kklxdm = "09"
			} else {
				return errors.New("课程类型错误")
			}

			// 设置NjdmId
			NjdmId := "20" + c.ClientBodyConfig.BhId[0:2]

			// 设置Xqm
			Xqm := cfg.Time.XueQi
			if Xqm == "1" {
				Xqm = "3"
			} else if Xqm == "2" {
				Xqm = "12"
			} else {
				return errors.New("学期格式错误")
			}

			// 获取doJxbId
			doJxbId, err := GetDoJxbId(c, v.KchID, v.JxbID, Kklxdm, NjdmId, cfg.Time.XueNian, Xqm)
			if err != nil {
				return err
			}

			// 选课
			if SelectFlag == "1" {
				err = SelectCourse(c, doJxbId, v.KchID, Kklxdm, v.Jxbzc)
				if err != nil {
					return err
				}
			} else {
				// 退课
				err = CancelCourse(c, doJxbId, v.KchID, cfg.Time.XueNian, Xqm)
				if err != nil {
					return err
				}
			}

			return nil
		}
	}
	return errors.New(CourseName + "课程不存在")
}

// KillCourse 选退课
func KillCourse(ctx context.Context, c *client.Client, cfg *config.Config, course *config.Course) {
	// 计算需要等待的时间
	// 时区
	loc, err := time.LoadLocation("Asia/Shanghai")
	if err != nil {
		log.Error("初始化时间地区失败，正在使用手动定义的时区信息 :", err)
		loc = time.FixedZone("CST", 8*3600)
	}
	t, err := time.ParseInLocation("2006-01-02 15:04:05", cfg.StartTime, loc)
	if err != nil {
		log.Error("时间格式错误: ", err)
		return
	}
	log.Info("选课开始时间: ", t)
	waitTime := t.Unix() - time.Now().Unix()

	select {
	case <-ctx.Done():
		return
	case <-time.After(time.Duration(waitTime) * time.Second):
		log.Info("时间已到，开始处理课程...")
		//go func() {
		for {
			select {
			case <-ctx.Done():
				return
			default:
				// 获取选课配置
				err = ReadClientBodyConfig(c)
				if err != nil {
					err = c.GetClientBodyConfig()
					if err != nil {
						log.Error("获取选课配置失败: ", err)
						continue
					}
				}
				// 保存选课配置
				if cfg.ClientBodyConfigEnabled == "1" {
					err = SaveClientBodyConfig(c)
					if err != nil {
						log.Error("保存选课配置失败: ", err)
						return
					}
				}
				log.Info("选课配置获取成功")
				// 选退课
				for k, v := range cfg.Course {
					// 处理课程
					log.Info("正在处理课程: ", k)
					err = HandleCourse(c, cfg, course, k, v)
					if err != nil {
						log.Error("处理课程失败: ", err)
						continue
					}
				}
				// 完成
				channel <- "完成"
				return
			}
		}
		//}()
	}
}

// SaveClientBodyConfig 保存选课配置
func SaveClientBodyConfig(c *client.Client) error {
	// 将c.ClientBodyConfig保存到文件CLientBodyConfig.json
	clientBodyConfig := c.ClientBodyConfig
	bytes, err := json.Marshal(clientBodyConfig)
	if err != nil {
		return err
	}

	err = os.WriteFile("ClientBodyConfig.json", bytes, 0666)
	if err != nil {
		return err
	}

	return nil
}

// ReadClientBodyConfig 读取选课配置
func ReadClientBodyConfig(c *client.Client) error {
	// 读取文件CLientBodyConfig.json到c.ClientBodyConfig
	bytes, err := os.ReadFile("ClientBodyConfig.json")
	if err != nil {
		return err
	}

	var clientBodyConfig client.ClientBodyConfig
	err = json.Unmarshal(bytes, &clientBodyConfig)
	if err != nil {
		return err
	}

	c.ClientBodyConfig = &clientBodyConfig

	return nil
}
